src/extensions/surveys.tsx

```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
         callSurveys(posthog, false)
     }, 3000)
 }
+
+const defaultBackgroundColor = '#eeeded'
+
+export function QuestionHeader({
+    question,
+    description,
+    backgroundColor,
+}: {
+    question: string
+    description?: string | null
+    backgroundColor?: string
+}) {
+    return (
+        <div style={{ backgroundColor: backgroundColor || defaultBackgroundColor }}>
+            <div className="survey-question">{question}</div>
+            {description && <div className="description" dangerouslySetInnerHTML={{ __html: description }} />}
+        </div>
+    )
+}
+
+export function Cancel({ onClick }: { onClick: () => void }) {
+    return (
+        <div className="cancel-btn-wrapper">
+            <button className="form-cancel" onClick={onClick}>
+                {cancelSVG}
+            </button>
+        </div>
+    )
+}
+
+export function BottomSection({
+    text,
+    submitDisabled,
+    appearance,
+    onSubmit,
+    link,
+}: {
+    text: string
+    submitDisabled: boolean
+    appearance: SurveyAppearance
+    onSubmit: () => void
+    link?: string | null
+}) {
+    const backgroundColorRef = Preact.createRef()
+    const [textColor, setTextColor] = useState('white')
+
+    useEffect(() => {
+        const color = getTextColor(backgroundColorRef.current)
+        setTextColor(color)
+    }, [appearance])
+
+    return (
+        <div className="bottom-section">
+            <div className="buttons">
+                <button
+                    className="form-submit"
+                    ref={backgroundColorRef}
+                    disabled={submitDisabled}
+                    type="button"
+                    style={{ color: textColor }}
+                    onClick={() => {
+                        if (link) {
+                            window.open(link)
+                        }
+                        onSubmit()
+                    }}
+                >
+                    {text}
+                </button>
+            </div>
+            {!appearance.whiteLabel && (
+                <PostHogLogo backgroundColor={appearance.backgroundColor || defaultBackgroundColor} />
+            )}
+        </div>
+    )
+}
+
+export function PostHogLogo({ backgroundColor }: { backgroundColor?: string }) {
+    const backgroundColorRef = Preact.createRef()
+    const [textColor, setTextColor] = useState('white')
+    useEffect(() => {
+        setTextColor(getTextColor(backgroundColorRef.current))
+    }, [])
```

Comment:
This block is almost identical across multiple components.

Could be turned into a custom hook that encapsulates it:

```tsx
export function useContrastingTextColor(options: { appearance: SurveyAppearance; defaultTextColor?: string }): {
    ref: React.RefObject<HTMLElement>
    textColor: string
} {
    const ref = useRef<HTMLElement>(null)
    const [textColor, setTextColor] = useState(options.defaultTextColor ?? 'white')

    useEffect(() => {
        if (ref.current) {
            const color = getTextColor(ref.current)
            setTextColor(color)
        }
    }, [options.appearance])

    return {
        ref,
        textColor,
    }
}


function MyComponent() {
      const { textColor, ref } = useContrastingTextColor({ appearance })
}


```